dbms/src/DataStreams/MergingSortedBlockInputStream.cpp

```diff
@@ -236,6 +236,21 @@ void MergingSortedBlockInputStream::merge(Block & merged_block, ColumnPlainPtrs
 			for (size_t i = 0; i < num_columns; ++i)
 				merged_columns[i]->insertFrom(*current->all_columns[i], current->pos);
 
+			if (row_sources)
+			{
+				/// Actually, current.impl->order stores source number (i.e. cursors[current.impl->order] == current.impl)
+				size_t source_num = 0;
+				size_t size = cursors.size();
+				for (; source_num < size; ++source_num)
+					if (&cursors[source_num] == current.impl)
+						break;
+
+				if (source_num != current.impl->order)
+					throw Exception("Developer's logical error");
```

Comment:
I plan to remove this check after B2B testing, since it makes extra loop over parts.